File: forward_backwards.py
import numpy as np
from scipy.stats import norm

def forward_backward(transition_probs, means, std, initial_dist, emissions, d):
    def emission_probs(x, state):
        return norm.pdf(x, loc=means[state], scale=std)
    forward_dists = forward(transition_probs, emission_probs, initial_dist, emissions, d)
    return forward_dists
    # backward_dists = backward(transition_probs, emission_probs, emissions)
    # print backward_dists
    # return normalize(np.multiply(forward_dists, backward_dists))

def backward(transition_probs, emission_probs, emissions):
    num_states = 2
    dist = normalize(np.ones((1, num_states)))
    dists = [dist]
    for emission in reversed(emissions): 
        dists.append(normalize(transition_probs * np.dot(emission_probs(emission), dist.T).T))
    dists.reverse()
    return np.row_stack(dists)

def forward(transition_probs, emission_probs, initial_dist, emissions, d):
    dists = [initial_dist] 
    for i, emission in enumerate(emissions):
        dist = []
        for st in xrange(d):
            if i == 0:
                prev_sum = dists[-1][st]
            else:
                prev_sum = sum([dists[-1][k]*transition_probs[k][st] for k in xrange(d)])
            dist.append(emission_probs(emission, st) * prev_sum)
        dists.append(normalize(dist))
    return dists

def normalize(array):
    return array / sum(array)

File: hmm-go/src/hmm/forward_backwards.go
package hmm

import "math"

func emission_probs(x float64, state int, p Parameters) float64 {
	normal_normalizer := 0.3989422804014327 / p.sigma
	return normal_normalizer * math.Exp(-1*(x-p.means[state])*(x-p.means[state])/(2*p.sigma*p.sigma))
}

func forward_backward(params Parameters) [][]float64 {
	return forward(params)
}

func forward(params Parameters) [][]float64 {
	dists := make([][]float64, len(params.y)+1)
	// var dists = [len(params.y) + 1][params.d]float64{}
	dists[0] = params.priors
	// var dist = [params.d]float64{}
	for j := 0; j < len(params.y); j++ {
		dist := make([]float64, params.d)
		for i := 0; i < params.d; i++ {
			// prev_sum := dists[j][i]
			dist[i] = emission_probs(params.y[j], i, params)
		}
		dists[j+1] = Normalize(dist)
	}
	return dists
}

func Normalize(input []float64) []float64 {
	sum := 0.0
	for i := 0; i < len(input); i++ {
		sum += input[i]
	}
	for i := 0; i < len(input); i++ {
		input[i] /= sum
	}
	return input
}
